src/java.base/share/classes/java/lang/ref/Finalizer.java

```diff
@@ -182,14 +192,16 @@ public void run() {
     }
 
     static {
-        ThreadGroup tg = Thread.currentThread().getThreadGroup();
-        for (ThreadGroup tgn = tg;
-             tgn != null;
-             tg = tgn, tgn = tg.getParent());
-        Thread finalizer = new FinalizerThread(tg);
-        finalizer.setPriority(Thread.MAX_PRIORITY - 2);
-        finalizer.setDaemon(true);
-        finalizer.start();
+        if (Holder.ENABLED) {
```

Comment:
Hello Stuart,
My understanding of the the lazy `Holder` is that it's there to delay the static initialization of whatever is part of that static block. In this case here, the `Holder` is being used within the `static` block of the `Finalizer` class, that too as the first thing. In this case, is that `Holder` class necessary?

Reply:
Huh, good catch! This was mostly left over from an earlier version of the flag that used system properties, which aren't initialized until after the Finalizer class is initialized.

It might be the case that the Holder can be removed at this point, since the finalization-enabled bit is no longer in a system property and is in a native class member that should be available before the VM is started.

I say "might" though because this occurs early in system startup, and weird things potentially happen. For example, suppose the first object with a finalizer is created before the Finalizer class is initialized. The VM will perform an upcall to Finalizer::register. An ordinary call to a static method will ensure the class is initialized before proceeding with the call, but this VM upcall is a special case.... I'll have to investigate this some more.

Reply:
@stuart-marks not sure I see how anything is different here compared to the existing logic. The `Finalizer` class is explicitly initialized quite early in the init process, but if a preceding class's initialization created an object with a finalizer then that same upcall would be involved.

Reply:
Do we even have to have a flag on Java side? It looks like these calls are only done as the upcalls from VM, so we might just keep the flag on VM side?

Reply:
@shipilev not sure what you mean by  "a flag on the Java side". The Java code just queries the VM for the finalization enabled/disabled state and uses that to control things.

Reply:
Yeah, "flag" is `Holder.ENABLED` here. I mean, are Java methods `registerFinalizer` and `runFinalization` called only by VM? If so, can VM check the whole thing on VM side, without going to Java and asking back from there?

Reply:
`registerFinalizer` does not expect to be called and only uses the "flag" as a form of assertion.

`runFinalization` is called from Java code.

Reply:
I think @shipilev asks a good question. This could be done completely in the VM without the changes to j.l.ref.Finalizer. The CLI option is for experimenting, at least in the short term, and should be benign to have the Finalizer thread running, it just won't do anything.

Reply:
Or, you could move the static initialization block that statrts the finalizer thread into the Finalizer.FinalizerThread class itself and then arrange for that class to be initialized explicitly immediately after the Finalizer class, but conditionally, only if the option to disable finalization was not specified...
This way the Finalizer class could still be initialized early, but the thread would not be started if it is not needed.

Reply:
The disable-finalization feature is a bit more than experimental. The goal is to provide a faithful representation of what the system will look like when finalization is removed. Of course most of that is objects' `finalize` methods not being called, but it also includes having no finalizer thread running, as well as having `runFinalization` (a public API) do nothing at all.

Reply:
@dholmes-ora If the Finalizer class is initialized explicitly and at the right time, then maybe we can do away with the Holder class entirely. Can you point me to where this is done?

Reply:
I pushed an update to remove the Holder class. It seems to continue to work fine. Thanks for pointing this out @jaikiran !

Reply:
Thank you Stuart, this changed version looks fine to me.